### ISSUE TITLE
Hoogling signatures with constraints

### ISSUE THREAD

**Issue by fredefox** [ndmitchell/hoogle#213]
I think it would be a nice feature if hoogle could show results for signatures with matching types but additional constraints. Say I hoogle

    (a -> b) -> f a -> f b

I would like for hoogle to suggest `fmap`. Also, it would be nice if hoogle was aware of subclass relationships so that the above would also be shown if I hoogle, say

    Applicative f => (a -> b) -> f a -> f b

Any thoughts?

**Comment by georgefst** [comment]
It seems in some cases [this already happens](https://hoogle.haskell.org/?hoogle=(a%20-%3E%20b)%20-%3E%20f%20a%20-%3E%20f%20b).

Anyway, I think it would be cool to allow syntax like `_ => (a -> b) -> f a -> f b`. Or even `_ f => (a -> b) -> f a -> f b` to only allow constraints on `f`. Although the latter is not valid in GHC.

**Comment by ndmitchell** [comment]
Agreed on all fronts - but it's hard. Someone will need to invest in looking at how it can work in the existing design.
